orttraining/orttraining/core/session/training_session.cc

```diff
@@ -188,11 +175,24 @@ Status TrainingSession::ConfigureForTraining(
         config.model_with_loss_function_path.value(), SaveOption::NO_RELOAD));
   }
 
+  // We need to get trainable weights to prevent constant folding from them. This works well if trainable weights are passed from config.
+  // For case we use GetTrainableModelInitializers to get trainable weights such as C++ frontend, it may get more initializers
+  // than trainable weights here as it's before transformers. So the constant folding may miss some nodes we actually can fold.
+  std::unordered_set<std::string> excluded_initializers =
+      !config.weight_names_to_train.empty()
+          ? config.weight_names_to_train
+          : GetTrainableModelInitializers(config.immutable_weights, loss_name);
+  for (const auto& weight_name_to_not_train : config.weight_names_to_not_train) {
+    excluded_initializers.erase(weight_name_to_not_train);
+  }
+  
+  ORT_RETURN_IF_ERROR(ApplyTransformationsToMainGraph(excluded_initializers));
+
   // derive actual set of weights to train
   std::unordered_set<std::string> weight_names_to_train =
       !config.weight_names_to_train.empty()
           ? config.weight_names_to_train
-          : GetTrainableModelInitializers(config.immutable_weights);
+          : GetTrainableModelInitializers(config.immutable_weights, loss_name);
   for (const auto& weight_name_to_not_train : config.weight_names_to_not_train) {
     weight_names_to_train.erase(weight_name_to_not_train);
```

Comment:
Similarly, can we add
```
Let's add
```
    LOGS(*session_logger_, INFO) << "Exclude "
                                 << weight_name_to_not_train
                                 << " from trainable list after applying graph transforms.";
```

Reply:
If there are many names, maybe some format like:
Excluding following weights from trainable list after applying graph transforms:
name0, name1, ...

---
In reply to: [434780997](https://github.com/microsoft/onnxruntime/pull/4116#discussion_r434780997) [](ancestors = 434780997)